### PR TITLE
[pylint] List-return-type checker

### LINKED ISSUE #3232: Port APIView checks to Pylint custom checks

### PR DESCRIPTION
**Pull request by l0lawrence**
Client methods that start with List_ should return an AsyncItemPaged/ItemPaged, or a custom class that then either inherits from ItemPaged or creates its own custom paging class with by_page()

Fixes #3232

### CODE REVIEW

File: tools/pylint-extensions/pylint-guidelines-checker/pylint_guidelines_checker.py [hunk 1]
```diff
@@ -875,8 +875,8 @@ class ClientListMethodsUseCorePaging(BaseChecker):
 
     def __init__(self, linter=None):
         super(ClientListMethodsUseCorePaging, self).__init__(linter)
-
-    def visit_functiondef(self, node):
+
```
Comment (catalinaperalta) [inline tools/pylint-extensions/pylint-guidelines-checker/pylint_guidelines_checker.py:878]: Nit: extra spaces

File: tools/pylint-extensions/pylint-guidelines-checker/tests/test_files/core_paging_acceptable.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+from azure.core.paging import ItemPaged
+
+class CosmosClient():
```
Comment (catalinaperalta) [inline tools/pylint-extensions/pylint-guidelines-checker/tests/test_files/core_paging_acceptable.py:3]: Just curious, was this copied from a real example?

File: tools/pylint-extensions/pylint-guidelines-checker/pylint_guidelines_checker.py [hunk 3]
```diff
@@ -885,23 +885,26 @@ def visit_functiondef(self, node):
...
+                    paging_class = False
+
                     try:
-                        # infer_call_result gives the method return value as a string
-                        returns = next(node.infer_call_result()).as_string()
-                        if returns.find("ItemPaged") == -1 and returns.find("AsyncItemPaged") == -1:
-                            self.add_message(
-                                msgid="client-list-methods-use-paging", node=node, confidence=None
-                            )
+                        if "def by_page" in next(node.value.infer()).as_string():
```
Comment (catalinaperalta) [inline tools/pylint-extensions/pylint-guidelines-checker/pylint_guidelines_checker.py:893]: You're now passing the return directly? And is this checking that both custom paging classes, ItemPaged and AsyncItemPaged define `by_page`?

File: tools/pylint-extensions/pylint-guidelines-checker/pylint_guidelines_checker.py [hunk 4]
```diff
@@ -875,8 +875,8 @@ class ClientListMethodsUseCorePaging(BaseChecker):
 
     def __init__(self, linter=None):
         super(ClientListMethodsUseCorePaging, self).__init__(linter)
-
-    def visit_functiondef(self, node):
+    
+    def visit_return(self, node):
```
Comment (catalinaperalta) [inline tools/pylint-extensions/pylint-guidelines-checker/pylint_guidelines_checker.py:879]: Assuming this change is to make the search more direct?

Comment (l0lawrence) [reply tools/pylint-extensions/pylint-guidelines-checker/pylint_guidelines_checker.py:879]: yes!

**Review by tjprescott — COMMENTED** [review]
Are there violations of this? Should this be added to the global CI suppressions for a time?

**Comment by l0lawrence** [comment]
> Are there violations of this? Should this be added to the global CI suppressions for a time?

My plan is to initially add them disabled and then see the impact/ give SDKs time to fix the warnings and then enable them.

**Review by tjprescott — APPROVED** [review]

**Review by catalinaperalta — APPROVED** [review]

**Pull request merged**